src/Servers/Kestrel/Core/src/Middleware/TlsListenerMiddleware.cs

```diff
@@ -0,0 +1,125 @@
+            // and no more data is coming, we can't block in a loop here because we will not get more data
+            if (buffer.Length < 6 && result.IsCompleted)
+            {
+                break;
+            }
+
+            var parseState = TryParseClientHello(buffer, out var clientHelloBytes);
+
+            // no data is consumed, it will be processed by the follow-up middlewares
+            input.AdvanceTo(buffer.Start);
+
+            switch (parseState)
+            {
+                case ClientHelloParseState.NotEnoughData:
+                    continue;
+
+                case ClientHelloParseState.NotTlsClientHello:
+                    await _next(connection);
+                    return;
+
+                case ClientHelloParseState.ValidTlsClientHello:
+                    _tlsClientHelloBytesCallback(connection, clientHelloBytes);
+                    await _next(connection);
+                    return;
+            }
+        }
+
+        await _next(connection);
+    }
+
+    private static ClientHelloParseState TryParseClientHello(ReadOnlySequence<byte> buffer, out ReadOnlySequence<byte> clientHelloBytes)
+    {
+        clientHelloBytes = default;
+
+        if (buffer.Length < 6)
+        {
+            return ClientHelloParseState.NotEnoughData;
+        }
+
+        var reader = new SequenceReader<byte>(buffer);
+
+        // Content type must be 0x16 for TLS Handshake
+        if (!reader.TryRead(out byte contentType) || contentType != 0x16)
+        {
+            return ClientHelloParseState.NotTlsClientHello;
+        }
+
+        // Protocol version
+        if (!reader.TryReadBigEndian(out short version) || IsValidProtocolVersion(version) == false)
+        {
+            return ClientHelloParseState.NotTlsClientHello;
+        }
+
+        // Record length
+        if (!reader.TryReadBigEndian(out short recordLength))
+        {
+            return ClientHelloParseState.NotTlsClientHello;
+        }
+
+        // 5 bytes are
+        // 1) Handshake (1 byte)
+        // 2) Protocol version (2 bytes)
+        // 3) Record length (2 bytes)
+        if (buffer.Length < 5 + recordLength)
+        {
+            return ClientHelloParseState.NotEnoughData;
+        }
+
+        // byte 6: handshake message type (must be 0x01 for ClientHello)
+        if (!reader.TryRead(out byte handshakeType) || handshakeType != 0x01)
+        {
+            return ClientHelloParseState.NotTlsClientHello;
+        }
+
+        clientHelloBytes = buffer.Slice(0, 5 + recordLength);
+        return ClientHelloParseState.ValidTlsClientHello;
+    }
+
+    private static bool IsValidProtocolVersion(short version)
+        => version == 0x0002  // SSL 2.0 (0x0002)
+        || version == 0x0300  // SSL 3.0 (0x0300)
+        || version == 0x0301  // TLS 1.0 (0x0301)
+        || version == 0x0302  // TLS 1.1 (0x0302)
+        || version == 0x0303  // TLS 1.2 (0x0303)
+        || version == 0x0304; // TLS 1.3 (0x0304)
```

Comment:
For Roslyn (C# compiler) it's better to write this as pattern-matching:
```suggestion
        => version is 0x0002  // SSL 2.0 (0x0002)
                   or 0x0300  // SSL 3.0 (0x0300)
                   or 0x0301  // TLS 1.0 (0x0301)
                   or 0x0302  // TLS 1.1 (0x0302)
                   or 0x0303  // TLS 1.2 (0x0303)
                   or 0x0304; // TLS 1.3 (0x0304)
```
Because [the compiler can play some tricks here](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA+ABATARgLABQhGAzAATZkDCZA3oWY2QA5QCWAbgIYAuMFOJGWAQIAGzIBJAM4A1LmLYATAApQIPCJDGyYUaWwgA7ABTSAFtB5kOeg8YCUDJi4C8APht3DRsq9dkAAwIgaFYjAD0EWQAyjEAMmRYAHSBZCbBoYFYTgQuLmhoXvo+fgGZJKGR0XGJJKnpFaG5+UyFxfa+/kEhlTjVZAAq8TFkOA0ZvYE4La1k7bYlxmU9gZXhZFFDI2PJ/ZNr2bOtC97L3U3km9HDo+PhB5Ukx/mnS13lUygA3Nfbd8lyI9AigXs5GKxOLx+BhBMJRBIZPJFKp1JptLp3jgzJYoNZFp0Xm5PATSmxpKswgNakkJpkwkS5nNoKtKmk/jT6mlgYFGUz8iymv0/rddtyhXz+UxBVMNltReN9k0cuCpa0ZYcrvKdvdGlNnqq1S4NZUfiKdYC9YdQYQAL5AA==), that reduce the number of branches needed.

And IMO pattern matching is a bit more legible.